Added data model for ContentSyncState

rhysyngsun opened the pull request:
#### Pre-Flight checklist

- [x] Migrations
  - [x] Migration is backwards-compatible with current production code
- [x] Testing
  - [x] Code is tested
  - [x] Changes have been manually tested

#### What are the relevant tickets?
Part of #209 

#### What's this PR do?
Adds the `ContentSyncState` data model and signals around `WebsiteContent` to ensure records are created.

#### How should this be manually tested?
The migrations should run and the app should start

mbertrand commented on content_sync/apps.py:
```diff
@@ -1,8 +1,12 @@
 
     name = "content_sync"
+
+    # def ready(self):
+    #     """ Application is ready """
+    #     import content_sync.signals  # pylint:disable=unused-import, import-outside-toplevel
```
Uncomment the above

mbertrand commented on content_sync/models.py:
```diff
@@ -1,4 +1,32 @@
 
-# Create your models here.
+
+class ContentSyncState(models.Model):
```
Do you think it's worth adding this to Django admin?

rhysyngsun replied:
I don't think there should be a need to work with this model in django-admin, since all of its fields shouldn't be modified outside of code.

mbertrand commented on content_sync/signals.py:
```diff
@@ -0,0 +1,18 @@
+    sender, instance, created, **kwargs
+):  # pylint: disable=unused-argument
+    """ Create/update the sync state """
+    api.upsert_content_sync_state(instance)
```
Do you think `ContentSyncState` objects should be backpopulated for all existing `WebsiteContent` objects via a management command or migration script?

mbertrand commented:
Also needs a formatting run to pass tests

mbertrand commented on websites/models.py:
```diff
@@ -99,11 +100,15 @@ def upload_file_to(self, filename):
 
+    def calculate_checksum(self) -> str:
+        """ Returns a calculated checksum of the content """
+        return sha256(str(self.markdown).encode("utf-8")).hexdigest()
```
Should include `metadata` too,  in case the markdown stays the same but the metadata changes.  For example, a `WebsiteContent` object of type `resource` will not have any markdown, just metadata.  Maybe something like this:

```python
sha256(json.dumps(self.metadata) + str(self.markdown).encode("utf-8")).hexdigest()
```

rhysyngsun commented:
@mbertrand I addressed your feedback, I also switched the model to `TimestampedModel` since that will probably be helpful, so you'll want to `manage.py migrate content_sync zero` before you pull the latest.

mbertrand commented on content_sync/migrations/0002_backpopulate_content_sync_state.py:
```diff
@@ -0,0 +1,36 @@
+    WebsiteContent = apps.get_model("websites", "WebsiteContent")
+    ContentSyncState = apps.get_model("content_sync", "ContentSyncState")
+
+    for content in WebsiteContent.objects.filter(content_sync_state=None):
```
There's about 88K of these on RC, I've got about the same amount in my local instance and docker memory usage went pretty high (~2.5 GB), wondering if that might cause problems on heroku.  One thing I did in a similar migration for open was use a chunked iterator:

```
def chunked_iterator(queryset):
    paginator = Paginator(queryset, 100)
    for page in range(1, paginator.num_pages + 1):
        for obj in paginator.page(page).object_list:
            yield obj
            
 for content in chunked_iterator(WebsiteContent.objects.filter(content_sync_state=None)):
 ```
 
Or can leave as is and see how it goes, maybe it's not necessary.

mbertrand reviewed (commented):
Looks great, one concern about memory usage in the migration.

mbertrand reviewed (approved):
👍
